### ISSUE TITLE
Preventing 301 redirects on URLs with no trailing slashes (Netlify)

### ISSUE THREAD

**Issue by lloydh** [gatsbyjs/gatsby#9207]
## Summary

URLs with no trailing slash on sites hosted by Netlify lead to an immediate 301 redirect to the page with a trailing slash.

``foo.com/bar`` --> ``foo.com/bar/``

This has a performance cost and implications for SEO.

_Is there a Netlify configuration that resolves these URLs without redirecting?_

## Relevant information

While this question is specific to Netlify, I did a quick review of other Gatsby sites featured in the Showcase and saw the same behaviour in many, but not all cases, for example:

[Hopper /company](https://www.hopper.com/company) - 301 redirect (Netlify)
[Impossible Foods /mission](https://impossiblefoods.com/mission) - 301 redirect (unknown)
[Cajun Bow Fishing /bows](https://cajunbowfishing.com/category/bows) - 301 redirect (Netllify)
[Braun /shavers-for-men](https://ca.braun.com/en-ca/male-grooming/shavers-for-men) - 200 no redirect  (unknown)

### Environment (if relevant)

Same behaviour in Gatsby v1 and v2.
I'm using [gatsby-plugin-remove-trailing-slashes](https://github.com/gatsbyjs/gatsby/tree/master/packages/gatsby-plugin-remove-trailing-slashes) and [gatsby-plugin-netlify](https://github.com/gatsbyjs/gatsby/tree/master/packages/gatsby-plugin-netlify).
Within the project all Links point to the non-trailing slash version.

**Comment by Yurickh** [comment]
I'm not really familiar on how Netlify runs the static site, but I know that this is the default behaviour for `express` when serving static folders. 
I'm not sure which are the performance costs you're implying here, can you link me to a reference? Would love to read more about that.

**Comment by lloydh** [comment]
The performance cost is the synchronous delay for the first byte of useful data caused by the redirect.

Using the Hopper example above, visiting [https://www.hopper.com/company](https://www.hopper.com/company) takes 150-300ms for the redirect, before any page data is received. On cellular connections with high latency it can add up to 1s.

**Comment by Yurickh** [comment]
I see. 

From my experience with static sites, we always redirect to the `/`-ending url whenever it represents a folder (with an implicit `index.html`), not a file (like `shavers-for-men.html`). If you don't, the static resolution will simply fail.

[This is also what the "pretty url" option on netlify does.](https://www.netlify.com/docs/redirects/#trailing-slash)

I guess the best approach here is to use the `/`-ending url as the canonical one, unless you have really solid reasons to do otherwise.

**Comment by lloydh** [comment]
@Yurickh I agree specifying the ``/``-ending urls as canonical is a pragmatic option but it does seem like advocating a schizophrenic url scheme for something that was easily solved with nginx / apache but not today's static site hosts.

If this really is the best approach then perhaps ``gatsby-plugin-remove-trailing-slashes`` should have this as an option (at least mentioned in the docs). Alternatively ``gatsby-plugin-canonical-urls`` could have an option for trailing slashes. I haven't found any other plugins that can set canonical meta tags.

I did come across #9025 but to be honest I'm surprised this hasn't been a bigger issue for a lot of folks given the popularity of ``/``-less Gatsby sites.

**Comment by kakadiadarpan** [comment]
@lloydh did you had a chance to look at this documentation of [Redirects|Netlify](https://www.netlify.com/docs/redirects/#trailing-slash)?

**Comment by luukdv** [comment]
Same issue here. Turned on 'Pretty URLs' in Netlify, but when I visit a page and remove the trailing slash in the address bar afterwards, I land on the non-trailing variant.

I agree with @lloydh about having an option (or by default?) in `gatsby-plugin-remove-trailing-slashes` to enforce a trailing slash. Right now it's not really a canonical, since the current `pathname` is used (https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby-plugin-canonical-urls/src/gatsby-browser.js#L9).

**Comment by Yurickh** [comment]
Just to clarify, the 'Pretty URLs' option in netlify WILL redirect you to the trailing slash variant:

> In addition to forwarding paths like /about to /about/ (a common practice in static sites and single page apps), it will also rewrite paths like /about.html to /about/.

**Comment by dimoFeeld** [comment]
We are haveing exactly the same problem. And we are now getting penelties from google. Is there a solution?

**Comment by lloydh** [comment]
> Just to clarify, the 'Pretty URLs' option in netlify WILL redirect you to the trailing slash variant

This is true but I haven't noticed any difference in behaviour with or without "Pretty URLs"; `/foo`  redirects to `/foo/` in either case. The desired behaviour is to rewrite _slashless_ urls to resolve without redirecting.

Making `/foo/` canonical is a workaround for the SEO penalties but AFAIK a full solution would only be possible in Netlify's routing layer… or by abandoning Netlify in favour of custom url rewriting …or by adopting _slash/_ urls. It's an unfortunate situation.

**Comment by abohannon** [comment]
I'm not having this issue with Netlify specifically, but my UTM query params are being deleted for this same reason. `/mypage?utm_source=google` becomes `/mypage/` and this is causing tracking issues.

**Comment by dja** [comment]
This is definitely still an issue, and we’ve experienced SEO penalties as well. Additionally, turning off Netlify’s pretty URLs feature seems to result in errors stating “Missing resources for /“ or “Missing resources for /slash/“. We’ve tried solutions recommended here: https://github.com/gatsbyjs/gatsby/issues/11524 but haven’t had any luck.

**Comment by 0505gonzalez** [comment]
Experiencing this issue as well.

**Comment by 0505gonzalez** [comment]
@dja Are you also on Netlify or are you using Github pages?

**Comment by dja** [comment]
We’re on Netlify.

Sent via Superhuman iOS ( https://sprh.mn/?vip=me@danieljacobarcher.com )

**Comment by 0505gonzalez** [comment]
@dja I'm on github pages, but the issue you're facing might be the same as mine. I've opened a new ticket and plan to create a PR shortly: https://github.com/gatsbyjs/gatsby/issues/12364

**Comment by 0505gonzalez** [comment]
TLDR: Github pages (and probably Netlify) add trailing forward slashes to folders. If you have something like `/public/somepage/index.html` and you visit `https://yourpage.com/somepage`, Github (and probably Netlify) will add a trailing slash because `somepage` is a directory.

**Comment by dja** [comment]
How do we have Gatsby create pages like `/public/somepage.html` instead of `/public/somepage/index.html` so that pages aren't within a namespaced directory?

Sent via Superhuman ( https://sprh.mn/?vip=me@danieljacobarcher.com )

**Comment by 0505gonzalez** [comment]
@dja That's what I'm currently trying to figure out in the other github issue I opened

**Comment by himynameistimli** [comment]
@0505gonzalez were you able to figure out a solution for this issue?

Just landed here as we're having the same problems with the url parameters getting lost in the redirect from the version without the trailing slash to the version with the trailing slash.

Only difference is that we're on S3 + Cloudfront.

We might look into using Lambda@Edge to handle the redirect unless we can figure out a way to get it to work with gatsby.

**Comment by 0505gonzalez** [comment]
@himynameistimli I did find a solution, proposed a code change in another thread. But seems like it might not be accepted so have not created a PR.

The gist of it:
* Hosting on github pages. Github follows the directory structure when serving files. E.g. if you hit `/somepage`, it will redirect to `/somepage/` because it's a directory (actual file is `/somepage/index.html`. 
* My proposed solution was that gatsby generate `/somepage.html` instead of `/somepage/index.html`

**Comment by garethgd** [comment]
Still experiencing this trailing slash redirect with or without Netlify's pretty URL option enabled.

**Comment by decimoseptimo** [comment]
@KyleAMathews 

> @himynameistimli I did find a solution, proposed a code change in another thread. But seems like it might not be accepted so have not created a PR.
> 
> The gist of it:
> 
> * Hosting on github pages. Github follows the directory structure when serving files. E.g. if you hit `/somepage`, it will redirect to `/somepage/` because it's a directory (actual file is `/somepage/index.html`.
> * My proposed solution was that gatsby generate `/somepage.html` instead of `/somepage/index.html`

@KyleAMathews this is seo danger
This subject isn't explained at all in https://www.gatsbyjs.org/docs/gatsby-link/
If someones decides to use no-trailing-slash-urls, a couple of days later the google serp becomes full of 301 redirections for your site.

**Comment by anhnk** [comment]
We are having the same problems as described in this PR. I'm just wondering if it is being actively worked on at the moment.

**Comment by wardpeet** [comment]
Not at this moment but we're open for Pull requests.

**Comment by statico** [comment]
I've just started evaluating Gatsby. Is there a way that, say for a given path `/foo`, have Gatsby generate `public/foo.html` instead of `public/foo/index.html`?

**Comment by decimoseptimo** [comment]
Not at the moment. The closest would be name your pages like `page.html.js`.
My advice to you would be to stick with trailing slashes.
For non-trailing slashes to work you'll need to do some server-side url rewriting.

**Comment by pewStocky** [comment]
Hi there, this would be great to have looked at as can cause issues SEO wise based on current builds etc.

One of the main things SEOs will look at when their internal dev team is wanting to use a new approach like Gatsby are things like these. 

Seems a bit weird that it forces / overwrites server side rules on trailing slashes, not sure how much work this will be but would help a lot! Thanks

**Comment by esetnik** [comment]
I just got stuck on the same issue. Is there any advantage to the current gatsby approach of using a directory structure for pages? Why can't we just render them all at the root as suggested by @statico?

**Comment by kimbaudi** [comment]
Netlify has pretty good documentation on handling custom redirects/rewrites.
For anyone interested in rewriting a path to prevent 301 redirects, take a look at [HTTP Status Codes - Redirect & Rewrite Rules](https://www.netlify.com/docs/redirects/#http-status-codes)

You should be able to create a `netlify.toml` file to specify your rewrite rules:

**sample netlify.toml with rewrite rule**

```
# Rewrite a path
/foo    public/foo.html    200
/bar/   /bar               200
```
I haven't tried this because I want "pretty URLs" that end with a trailing slash.
The only downside to this approach is that it is very tedious and error prone to specify all paths in your website to be rewritten without trailing slash :disappointed: 

Also, this issue seems to be related to https://github.com/gatsbyjs/gatsby/issues/15317

**Comment by esetnik** [comment]
Would it be as easy as modifying https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby/src/utils/worker/render-html.js#L10

**Comment by kimbaudi** [comment]
I don't know about modifying https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby/src/utils/worker/render-html.js#L10.

I just wanted to mention that there is a way to do it in Netlify using `netlify.toml`

**Comment by agoldis** [comment]
FYI the same issue happens when serving gatsby artifacts on S3 behind CloudFront (prob that's what Netlify do behind the scenes)

**Comment by wardpeet** [comment]
Is it possible to add a reproduction repo that we can build on netlify or wherever to mimic this problem and get it resolved

**Comment by bijenkorf-james-wakefield** [comment]
I don't think that this issue is related to Netlify or Github pages. I can reproduce it locally running `gatsby serve`, which appears to be [expected behaviour Gatsby's redirect handling](https://github.com/gatsbyjs/gatsby/blob/561d33e2e491d3971cb2a404eec9705a5a493602/packages/gatsby/src/redux/reducers/__tests__/redirects.js#L4).

@lloydh / @luukdv  - can you reproduce locally?

**Comment by dougwithseismic** [comment]
This is a pretty important issue for us, too.  You can see this behaviour even with https://gatsbyjs.github.io/gatsby-starter-default

**Comment by pewStocky** [comment]
Hey doug, i'm currently looking into this too with my dev team (i'm in-house SEO) - my post here: https://github.com/gatsbyjs/gatsby/issues/15317 

Will update you / this thread if we come up with any solutions etc - cheers

**Comment by dougwithseismic** [comment]
@jaapstronks Are you by any chance running with netlify?  https://github.com/gatsbyjs/gatsby/issues/15317#issuecomment-530048373

**Comment by PolGuixe** [comment]
Can someone summarise the current situation?

**Solution:** still WIP. 

**Hack:** 

1. Ensure trailing slashes are used in all routes
1. Add a trailing slash to the canonical link

Please feel free to correct me and I'll amend/delete the above.

**Comment by brettmstevens7** [comment]
If you're using Gatsby + Netlify, here's how we solved it.

First, any pages in the `/pages` directory, which are created automatically by Gatsby, will have a trailing slash. You can see this behavior by entering a URL without a trailing slash directly into the search bar. The network tab will show a 301 as the page is redirected to the one with the trailing slash.

Try running this GraphQL query and you will see that the path for these pages has a trailing slash. 

```
query MyQuery {
  allSitePage {
    edges {
      node {
        path
      }
    }
  }
}
```

To remove the trailing slash from the page path, you can add ![this code snippet](https://www.gatsbyjs.org/docs/creating-and-modifying-pages/#removing-trailing-slashes) from the Gatsby docs to your `gatsby-node.js` file:

```
// Replacing '/' would result in empty string which is invalid
const replacePath = path => (path === `/` ? path : path.replace(/\/$/, ``))
// Implement the Gatsby API “onCreatePage”. This is
// called after every page is created.
exports.onCreatePage = ({ page, actions }) => {
  const { createPage, deletePage } = actions

  const oldPage = Object.assign({}, page)
  // Remove trailing slash unless page is /
  page.path = replacePath(page.path)
  if (page.path !== oldPage.path) {
    // Replace new page with old page
    deletePage(oldPage)
    createPage(page)
  }
}
```

Next, you need to turn off the pretty URL setting in Netlify. If you have Asset optimization disabled, enable it, turn off pretty URLs, and then disable it again.

<img width="1239" alt="netlify-setting" src="https://user-images.githubusercontent.com/27828739/68625176-58817380-048d-11ea-9cf1-31473620bd08.png">

Push a new build and then observe the network requests for a page that had a trailing slash. For instance, after typing `https://www.software.com/code-time` then pressing `Enter`, the first request is a 200 (there isn't a 301 like before):

<img width="1166" alt="200-resp" src="https://user-images.githubusercontent.com/27828739/68625282-9a121e80-048d-11ea-86da-82662c386eb2.png">

If anyone has any better solutions, let me know! We haven't solved the trailing slash on `localhost`.

**Comment by AAverin** [comment]
@wardpeet @KyleAMathews Would like to bring your attention to this issue again because it's a major problem for SEO, as many people in the tread already mentioned.
Asking for PRs is great, but this particular one looks to be related to some of the core functionalities of the Gatsby engine, so I myself wouldn't take responsibility touching it, not knowing all ins and outs of Gatsby.

Could you or someone with deep knowledge take a look and offer some kind of solution?

**Comment by lawwantsin** [comment]
This is a problem in AWS as well.  Gatsby is a scam.

**Comment by dougwithseismic** [comment]
Guys, it's a setting within Netlify as described above. Perhaps a new issue
needs to be created.

**Comment by dospolov** [comment]
Two years later - still an issue

**Comment by stldo** [comment]
I made a plugin from a code that I normally use in websites hosted at Netlify. It creates a `.html` file for each page, which disables the 301 redirect for paths without trailing slashes. It works very well with simple websites, I hope it helps someone. More info can be found on the [plugin page](https://github.com/stldo/gatsby-plugin-create-page-html).

**Comment by ayZagen** [comment]
this also happens with nginx server.

**Comment by lyxious** [comment]
> this also happens with nginx server.
> 
> **EDIT**: this issue has nothing with gatsby. it is web server misconfiguration. I have checked output files and it seems gatsby creates a directory for each page and and index.html in it. So I had to change my nginx url resolving as following:
> 
> ```nginx
> location / {
>   try_files $uri $uri/index.html $uri.html =404;
> }
> ```
> 
> The `$uri/index.html` is resolving correct file without redirect. If that doesn't exists or it is `$uri/` ( most nginx conf examples uses that ) it will create a redirect with trailing slash. It is also stated in nginx [documantation](http://nginx.org/en/docs/http/ngx_http_core_module.html#location).
> 
> > In response to a request with URI equal to this string, but without the trailing slash, a permanent redirect with the code 301 will be returned to the requested URI with the slash appended.
> 
> I have not used Netlify but I believe same thing would apply to it.

This is not the case, even if you have a blank nginx config, any attempt to access a valid directory will result in a 301 with the trailing slash. It's a common misconception that it is related to the ```try_files```.
The documentation is also poor surrounding this, the [location](http://nginx.org/en/docs/http/ngx_http_core_module.html#location) documentation makes it seem like the 301 redirect is only for routes that are proxied.

**Comment by atkinson** [comment]
This is a bug in the Netlify UI.

Here's a fix: https://community.netlify.com/t/remove-trailing-slash-redirect-for-gatsby-gatsby-cloud-netlify-website/20976/8

**Comment by ayZagen** [comment]
> This is not the case, even if you have a blank nginx config, any attempt to access a valid directory will result in a 301 with the trailing slash. It's a common misconception that it is related to the `try_files`.
> The documentation is also poor surrounding this, the [location](http://nginx.org/en/docs/http/ngx_http_core_module.html#location) documentation makes it seem like the 301 redirect is only for routes that are proxied.

There is nothing said about `try_files` it is about how file resolving works in nginx. `try_files` is just a helper directive. That code is just an example to show my usage and solution. I agree with you about documentation being poor. The redirection to a directory with a slash will be performed by an undocumented module named `ngx_http_static_module`. If you want to disable that behaviour you need to compile nginx yourself. I believe no one here would try this hard for it.

**Comment by lyxious** [comment]
> > This is not the case, even if you have a blank nginx config, any attempt to access a valid directory will result in a 301 with the trailing slash. It's a common misconception that it is related to the `try_files`.
> > The documentation is also poor surrounding this, the [location](http://nginx.org/en/docs/http/ngx_http_core_module.html#location) documentation makes it seem like the 301 redirect is only for routes that are proxied.
> 
> There is nothing said about `try_files` it is about how file resolving works in nginx. `try_files` is just a helper directive. That code is just an example to show my usage and solution. I agree with you about documentation being poor. The redirection to a directory with a slash will be performed by an undocumented module named `ngx_http_static_module`. If you want to disable that behaviour you need to compile nginx yourself. I believe no one here would try this hard for it.

OP's question is about why ```foo.com/bar``` redirects to ```foo.com/bar/```. Your initial answer suggests it has to do with the ```try_files``` config, but it doesn't. The [documentation](http://nginx.org/en/docs/http/ngx_http_core_module.html#location) you provided also has nothing to do with why this request returns a 301. The location documentation, with respect to 301, is in regards to specific ```*_pass``` processes which is independant of ```try_files```.

**Comment by wardpeet** [comment]
Hey, sorry I don't have an update right now but I'll read over this thread again and see if I can get some action items out of it and create a task list so y'all can help us fix these issues 🙏

**Comment by mlenser** [comment]
> This is a bug in the Netlify UI.
> 
> Here's a fix: https://community.netlify.com/t/remove-trailing-slash-redirect-for-gatsby-gatsby-cloud-netlify-website/20976/8

This is indeed the case.

Here is how your netlify config should look like:
![image](https://user-images.githubusercontent.com/235857/93735929-e9ad9480-fbde-11ea-9870-a4cb7568d227.png)

Disabling optimization at the top level apparently turns on the pretty URLs, even though it visually looks like that isn't the case:
![image](https://user-images.githubusercontent.com/235857/93735960-ff22be80-fbde-11ea-86f6-e2df7202d67d.png)

So don't check the checkbox next to "Disable asset optimization"

**Comment by flackjap** [comment]
> > This is a bug in the Netlify UI.
> > Here's a fix: https://community.netlify.com/t/remove-trailing-slash-redirect-for-gatsby-gatsby-cloud-netlify-website/20976/8
> 
> This is indeed the case.
> 
> Here is how your netlify config should look like:
> ![image](https://user-images.githubusercontent.com/235857/93735929-e9ad9480-fbde-11ea-9870-a4cb7568d227.png)
> 
> Disabling optimization at the top level apparently turns on the pretty URLs, even though it visually looks like that isn't the case:
> ![image](https://user-images.githubusercontent.com/235857/93735960-ff22be80-fbde-11ea-86f6-e2df7202d67d.png)
> 
> So don't check the checkbox next to "Disable asset optimization"

I've just lost 3 hours because of this.

BOLEST.

**Comment by alvinometric** [comment]
@jlengstorf Do you know if this is intended behaviour ?

**Comment by sohamsj** [comment]
slash

**Comment by jlengstorf** [comment]
@alvinometric I'm not sure — I've sent this over to our UI team for review. it does look like if this isn't a bug, it could do with some clarification

**Comment by leomelzer** [comment]
Thanks for all the helpful comments which lead us in the right direction. 

If it still doesn't work after setting it to @mlenser's comment (https://github.com/gatsbyjs/gatsby/issues/9207#issuecomment-695913229), make sure to check your `netlify.toml` for pretty_urls.

Settings in the toml take precedence, see the docs: https://docs.netlify.com/configure-builds/file-based-configuration/#deploy-contexts

> UI settings are overridden if a netlify.toml file is present in the root folder of the repo and there exists a setting for the same property/redirect/header in the toml file.

**Comment by LekoArts** [comment]
Since multiple people have reported this as a bug in Netlify's UI / the behavior being the result of a misconfigured hosting, I'll close this one here as resolved (and not an issue with Gatsby). Please follow the linked issues to see how/when it's resolved. Thanks for providing your context and solutions here for future Google users (hello 👋 ).

**Comment by jon-sully** [comment]
Hey 👋🏻

I know this issue lapsed and got closed but I really think it's important to recap on a couple of things here. The impetus for this issue is that **a)** there's a disconnect between Gatsby's routing defaults and Netlify's routing configurations, and **b)** there are _serious_ SEO penalties in play if a Gatsby site doesn't have the trailing-slash / no-trailing-slash issue solved, since a site serving the same content on both URLs (duplicate content) gets knocked on SEO pretty hard. Technically this isn't Gatsby's fault, but as it pertains to all Gatsby users hosting on Netlify, it does seem like a major issue.. or a major risk at the very least. 

Solving this problem by disabling "Pretty URLs" in the (yes, awfully borked / painful UX'd) Netlify Asset Optimization panel can open your site up to the duplicate content issue since content may be available at both the un-slashed and the slashed version of your URL path. It's important too to note that if a Gatsby site is available on both `/test` _and_ `/test/` but 'fixes itself', you may just be seeing the Gatsby runtime adjust your address bar via the Browser History API - the **super** important part has nothing to do with what happens when Gatsby actually runs in the browser - it's the part where Netlify is serving the same content on multiple URLs - the slash and the non-slash paths. 

This _is_ fixable and there _is_ a way to get everything working smoothly and on a unified path / slash structure, but it's not disabling 'Pretty URLs'. The tl;dr: is that Netlify really works best / has biases toward _using_ the trailing slash, and unified content pathing on Netlify requires the trailing slash. I go into the how-to and whatnot here: 

https://jonsully.net/blog/trailing-slashes-on-netlify/

But I would definitely urge folks to carefully check (from a CLI HTTP tool preferably) which paths (slash and/or no-slash) are resolving to their content on their sites. If both the slash and no-slash paths are resolving to your content, your SEO will hurt for it.

Hope that helps 😕 I know it's not exactly fun news and goes a tad against the grain of this thread thus far, but I was experiencing similar issues with my Gatsby site and showing analytics for both `/about` and `/about/` (in my case) which lead me down this rabbit hole. Hopefully it helps someone!

**Comment by yanneves** [comment]
Given gatsbyjs.com itself resolves HTTP 200 with or without trailing slash (duplicate content), it seems this issue is a bit of an afterthought.

```bash
$ curl -I https://www.gatsbyjs.com/plugins/
HTTP/2 200
```

```bash
$ curl -I https://www.gatsbyjs.com/plugins
HTTP/2 200
```

I noticed this issue like others here when I migrated from WordPress to Gatsby, where the previous WordPress configuration stripped trailing slashes. The only way I can see to avoid duplicate content is to cave to the 301 redirects and introduce trailing slashes. Is the SEO penalty for redirecting existing pages like this still relevant? It could be this is a lingering idea in SEO that no longer matters. But otherwise it's a dangerous assumption in Gatsby.

This appears to be baked into the directory structure of the static generated site:

```
public/
├── index.html
├── some-other-page
│   └── index.html
└── some-page
    └── index.html
```

A browser would interpret that as `example.com/some-page/` and it would be hacky to force it to remove the trailing slash. Generated content would need to respect a configuration option to instead output the following when we want to remove trailing slash:

```
public/
├── index.html
├── some-other-page.html
└── some-page.html
```

The above would then be interpreted by the browser as `example.com/some-page`.

Do we know if Gatsby core is strictly expecting a directory structure for pages? There may be assumptions elsewhere that these are always output as directories. If we can identify those assumptions (or the lack of) configuring a non-trailing slash output like above would solve this issue.

**Comment by krzysieqq** [comment]
Any updates?

**Comment by MakowskiHubert** [comment]
The same problem with the nginx server and solved by https://github.com/gatsbyjs/gatsby/issues/9207#issuecomment-681870472

**Comment by hazem3500** [comment]
If you want your website to not have any trailing slash and also work with Netlify you can use the [`gatsby-plugin-netlify`](https://github.com/gatsbyjs/gatsby/tree/master/packages/gatsby-plugin-netlify) and in `gatsby-node.js` add

```js
const replacePath = path => (path === `/` ? path : path.replace(/\/$/, ``))

exports.onCreatePage = ({ page, actions }) => {
  const { createRedirect } = actions
  if(!page.path.includes('.html') && page.path !== '/') {
    createRedirect({ fromPath: `${page.path}/`, toPath: page.path, isPermanent: true })
  }
}
```

this will 301 redirects to all trailing slash paths.
Note if you are also using `createPages` Gatsby node API you'll need to add it there also 
```js
exports.createPages = async ({ actions, graphql }) => {
    const { createPage, createRedirect } = actions
       // ...
       pages.forEach(page => {
          // ...
          createRedirect({ fromPath: `${page.path}/`, toPath: page.path, isPermanent: true })
       })
    })
}
```

**Comment by WhiteHoodHacker** [comment]
By disabling "Pretty URLs" in Netlify and ending up with duplicate content at trailing-slash URLs and non-trailing-slash URLs, wouldn't a simple fix be to add a `<link>` tag with the canonical URL using the preferred scheme? Plenty of sites serve duplicate content fine with this.

**Comment by wojtekidd** [comment]
> I'm not having this issue with Netlify specifically, but my UTM query params are being deleted for this same reason. `/mypage?utm_source=google` becomes `/mypage/` and this is causing tracking issues.

Is anyone else having this problem still? There has to be a solution for a self-hosted gatsby to somehow accept utm's? 
I tried to do it this way in `gatsby-node.js`:

```
createRedirect({ fromPath: '/banana', toPath: '/shop?utm_source=banana&utm_medium=podcast&utm_campaign=banana', isPermanent: true });
```
